### DISCUSSION TITLE
Is FormProvider still have performance problem?

### DISCUSSION THREAD

**Discussion by genshinw** [react-hook-form/react-hook-form#8747] [Q&A] [answered]
Is FormProvider still have performance problem? 
I have not encountered  similar performance issues like https://react-hook-form.com/advanced-usage#FormProviderPerformance

Whether it is not recommended use it?

**Comment by Moshyfawn** [comment, marked as answer]
`FormProvider` never had "performance problems". The way React tree re-renders due to state updates passed via React `Context` is natural to the way React Context works with `Provider` + `Consumer`.

It's always recommended to use it responsibly in cases, where your form components tree is quite nested and passing form methods props by hand / via props is just unproductive _(same goes every other React Context usage in React eco-system)_

**Reply by genshinw** [reply]
Oh, Thanks, I may got something wrong  after read the post https://react-hook-form.com/advanced-usage#FormProviderPerformance.
I also want ask another question, how to write the strict generic type such as `Controller` without pass the control props when use the FormProvider ?
